p2p/transport/webrtc/stream_write.go

```diff
@@ -0,0 +1,172 @@
+package libp2pwebrtc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+	"github.com/pion/datachannel"
+	"google.golang.org/protobuf/proto"
+)
+
+func (s *webRTCStream) Write(b []byte) (int, error) {
+	if !s.stateHandler.AllowWrite() {
+		return 0, io.ErrClosedPipe
+	}
+
+	// Check if there is any message on the wire. This is used for control
+	// messages only when the read side of the stream is closed
+	if s.stateHandler.State() == stateReadClosed {
+		s.readLoopOnce.Do(func() {
+			go func() {
+				// zero the read deadline, so read call only returns
+				// when the underlying datachannel closes or there is
+				// a message on the channel
+				s.rwc.(*datachannel.DataChannel).SetReadDeadline(time.Time{})
+				var msg pb.Message
+				for {
+					if s.stateHandler.Closed() {
+						return
+					}
+					err := s.readMessageFromDataChannel(&msg)
+					if err != nil {
+						if errors.Is(err, io.EOF) {
+							s.close(true, true)
+						}
+						return
+					}
+					if msg.Flag != nil {
+						state, reset := s.stateHandler.HandleInboundFlag(msg.GetFlag())
+						if state == stateClosed {
+							log.Debug("closing: after handle inbound flag")
+							s.close(reset, true)
+						}
+					}
+				}
+			}()
+		})
+	}
+
+	const chunkSize = maxMessageSize - protoOverhead - varintOverhead
+
+	var (
+		err error
+		n   int
+	)
+
+	for len(b) > 0 {
+		end := len(b)
+		if chunkSize < end {
+			end = chunkSize
+		}
+
+		err := s.writeMessage(&pb.Message{Message: b[:end]})
+		n += end
+		b = b[end:]
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, err
+}
+
+func (s *webRTCStream) writeMessage(msg *pb.Message) error {
+	var writeDeadlineTimer *time.Timer
+	defer func() {
+		if writeDeadlineTimer != nil {
+			writeDeadlineTimer.Stop()
+		}
+	}()
+
+	for {
+		if !s.stateHandler.AllowWrite() {
+			return io.ErrClosedPipe
+		}
+		// prepare waiting for writeAvailable signal
+		// if write is blocked
+		deadlineUpdated := s.writerDeadlineUpdated.Wait()
+		writeAvailable := s.writeAvailable.Wait()
+
+		writeDeadline, hasWriteDeadline := s.getWriteDeadline()
+		if !hasWriteDeadline {
+			writeDeadline = time.Unix(math.MaxInt64, 0)
```

Comment:
No, point is to have a timer that never finishes in case no deadline, this way we can use always the timer, even if we do not want deadline. Other option would be to duplicate the select statement, with once the deadline and once ot. A zero time would make a timer that goes immediately off...